xdsl/pattern_rewriter.py

```diff
@@ -318,11 +319,19 @@ class RewritePattern(ABC):
     A side-effect free rewrite pattern matching on a DAG.
     """
 
+    traits: ClassVar[tuple[type[Trait], ...]] = ()
+
+    def _match_and_rewrite(self, op: Operation,
```

Comment:
Why not call this `satisfies_trait_constraints`?

Maybe we should have a "two stage" pattern application, one that checks `is_applicable` and one that does `apply_rewrite`, that way we split these two things into their respective methods and one can overwrite/extend one without touching the other.

Reply:
Because in a follow-up branch it calls the more strictly typed function. You're right that there could be a separate function that tests for trait conformance and type, and returns a boolean. I'll take a look at exploring that.

Reply:
While this could work, this is sometimes annoying, because that would mean that the operation/traits you match in the `is_applicable` section are not in scope anymore in the `apply_rewrite` section.